include/deal.II/base/discrete_time.h

```diff
@@ -0,0 +1,190 @@
+#define dealii_discrete_time_h
+
+#include <deal.II/base/config.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * Provides a means to keep track of the simulation time in a time-dependent
+ * simulation. It manages stepping forward from an initial time to a final
+ * time. It also allows adjusting the time step size during the simulation.
+ * It is guaranteed that at all times the current simulation time is in the
+ * closed interval between the start time and the end time.
+ *
+ * You can loop over all of the time steps by using a for loop
+ * @code
+ *   for (DiscreteTime time(0., 1., 0.1);
+ *        time.get_current_time() != time.get_end_time();
+ *        time.advance_time())
+ *   {
+ *     // Insert simulation code here
+ *   }
+ * @code
+ *
+ * @author Reza Rastak, 2019
+ */
+class DiscreteTime
+{
+public:
+  /**
+   * Constructor
+   */
+  DiscreteTime(const double start_time,
+               const double end_time,
+               const double start_step_size);
+
+  /**
+   * Return the current time.
+   */
+  double
+  get_current_time() const;
+
+  /**
+   * Return the start time.
+   */
+  double
+  get_start_time() const;
+
+  /**
+   * Return the end of the time interval.
+   * The final time step ends exactly at this point. This exact floating-point
+   * equality is very important because it allows us to use the expression
+   * <code>time.get_current_time() != time.get_end_time()</code> as the
+   * conditional statement in a for loop to check if the end time is reached.
+   */
+  double
+  get_end_time() const;
+
+  /**
+   * Return the size of the step from current time step to the next.
+   */
+  double
+  get_next_step_size() const;
+
+  /**
+   * Set the value of the next time step size. The next time advance_time()
+   * is called, the newly set @p time_step_size will be used to advance
+   * the simulation time. However, if the step is too large such that the next
+   * simulation time exceeds the end time, the step size is truncated.
+   * Additionally, if the step size is such that the next simulation time
+   * approximates the end time, the step size is adjusted such that the
+   * next simulation time exactly matches the end time.
+   */
+  void
+  set_next_step_size(const double time_step_size);
+
+  /**
+   * Advance the current time based on the value of the current step.
+   * If we are already at the end time, this function does nothing.
+   * Otherwise if the current time is less than the end time, the time
+   * is advanced to the next simulation time. If you want to adjust the
+   * next simulation time, call the method set_next_step_size() before
+   * calling this method. If you call this function repeatedly, the time
+   * is increases with the same step size until it reaches the end
+   * time. See the documentation of set_next_step_size() for explanation
+   * of the rules for automatic adjustment of the step size.
+   *
+   * @pre Current time must be smaller than the end time. The object cannot
```

Comment:
That's in contradiction to the statement above where you say that if you are already at the end time, nothing happens.